### PR TITLE
feat: handle Gateway disconnections

### PR DESCRIPTION
**Pull request by EmreTech**
This PR adds support for shards/the shard manager to handle disconnections. To do this, the shard manager has a loop that is started in the connect function that actively monitors every shard. If a shard has disconnected and is able to reconnect, then the shard's connect function is called. Otherwise, the shard is recreated.

### CODE REVIEW

**Review by plun1331 — APPROVED** [review]
lgtm

File: pycord/internal/gateway/shard.py [hunk 1]
```diff
@@ -26,7 +27,7 @@
 from typing import Any
 
 from aiohttp import WSMsgType
-from discord_typings.gateway import HelloEvent
+from discord_typings.gateway import GatewayEvent#HelloEvent, InvalidSessionEvent
```
Comment (VincentRPS) [inline pycord/internal/gateway/shard.py:30]:
```suggestion
from discord_typings.gateway import GatewayEvent, HelloEvent, InvalidSessionEvent
```

**Comment by EmreTech** [comment]
There is a current problem that Vincent has pointed out: because of the `while True` loop in the shard manager loop, it seems to take control over everything else in the event loop, which is bad. But there isn't another good solution that keeps in this loop but doesn't block the event loop.

**Comment by EmreTech** [comment]
> There is a current problem that Vincent has pointed out: because of the `while True` loop in the shard manager loop, it seems to take control over everything else in the event loop, which is bad. But there isn't another good solution that keeps in this loop but doesn't block the event loop.

Problem resolved.

File: pycord/internal/gateway/shard.py [hunk 2]
```diff
@@ -91,27 +98,37 @@ def __init__(
...
-        self._rot_done.set()
+            self._rot_done.set()
 
-        try:
-            await asyncio.create_task(self.start_clock())
-        except:
-            # user has most likely existed this process.
-            return
+            try:
+                await asyncio.create_task(self.start_clock())
```
[the remark this answers is not in the document]
Comment (EmreTech) [reply pycord/internal/gateway/shard.py:126]: Vincent wrote the start_clock function, so ask him

[the remark this answers is not in the document]
Comment (plun1331) [reply pycord/internal/gateway/shard.py:126]: @VincentRPS

**Review by VincentRPS — APPROVED** [review]

**Pull request merged**